Add solution 7 (#1707)
graphql/graphql-wg: rfcs/SemanticNullability.md

published), allowing for the "partial success" feature of GraphQL to continue to
shine and not compromising the resiliency of legacy deployed app versions.

| [1][solution-1] | [2][solution-2] | [3][solution-3] | [4][solution-4] | [5][solution-5] | [6][solution-6] | [7][solution-7] |
| --------------- | --------------- | --------------- | --------------- |-----------------|-----------------|-------------|

> | ✅              | ✅              | ✅              | ✅              | 🚫              | ✅          | ✅            |

Reviewer: ```suggestion
| ✅              | ✅              | ✅              | ✅              | 🚫              | ✅          | ⚠️            |
```

With `propagateErrorOnAllNonNullFields: true`, adding `!` (semantic non-null) in existing positions will change error propagation boundaries in existing (deployed) executable documents, in the same way that it would for solution 6, so semantic non-null could not be added until `propagateErrorOnAllNonNullFields` is set to `false` (i.e. after all clients and tooling migrate).

We do not want to break existing tooling.

| [1][solution-1] | [2][solution-2] | [3][solution-3] | [4][solution-4] | [5][solution-5] | [6][solution-6] | [7][solution-7] |
|-----------------|-----------------|-----------------|-----------------|-----------------|-----------------|-------------|

> | ✅               | ❔              | ✅               | ❔               | ✅               | ✅          | ✅            |

Reviewer: ```suggestion
| ✅               | ❔              | ✅               | ❔               | ✅               | ✅          | ⚠️            |
```

This solution is non-breaking _if_ you keep `propagateErrorOnAllNonNullFields: true` enabled forever.

Otherwise, it is a breaking change as it changes the meaning of `NON_NULL` in the schema, preventing adoption until all clients and tooling have been updated.

> allow inconsequential changes in behavior, but bubbling the error up isn't
> inconsequential.)

| [1][solution-1] | [2][solution-2] | [3][solution-3] | [4][solution-4] | [5][solution-5] | [6][solution-6] | [7][solution-7] |
|-----------------|-----------------|-----------------|-----------------|-----------------|-----------------|-------------|

> | ✅               | ❔              | ✅               | ✅               | 🚫               | ✅          | ✅            |

Reviewer: ```suggestion
| ✅               | ❔              | ✅               | ✅               | 🚫               | ✅          | ⚠️            |
```

Unless `propagateErrorOnAllNonNullFields` is set to `true`, any existing executable documents that introspect types will receive a change in meaning for `NON_NULL`.

> an outcome of this whole effort I’d like to see happening.

| [1][solution-1] | [2][solution-2] | [3][solution-3] | [4][solution-4] | [5][solution-5] | [6][solution-6] | [7][solution-7] |
|-----------------|-----------------|-----------------|-----------------|-----------------|-----------------|-------------|

> | ✅               | ✅              | ✅               | ✅               | 🚫               | ⚠️           | ✅            |

Reviewer: If `propagateErrorOnAllNonNullFields: true` is set, then adding `!` to more fields changes the error boundaries, and thus means that existing users cannot add `!` in more places without breaking existing clients error resilience.

Setting `propagateErrorOnAllNonNullFields: false`, even once all the fields are marked up with `@propagateError`, is a breaking change from introspection point of view so cannot be done until all clients and tooling have been updated to understand the new `NON_NULL` meaning and the additional introspection field. As such, `propagateErrorOnAllNonNullFields: false` cannot be safely adopted in a public API, partner API, or any other API where you do not control all clients and tooling.

Thus I'd argue that this solution fails P except for private APIs (and even there, the migration story is painful if you don't have teams dedicated to it).

Author: Updated in https://github.com/graphql/graphql-wg/pull/1707/commits/a85fbab15a56ca5684f3f6bfd95d1f4bf54c1a5c

  - ✅
- [R][criteria-r]
  - ✅ Directive is only valid on output positions.

## 💡 7. `@propagateError` directive

[solution-7]: #-7-propagateerror-directive

**Champion**: @leebyron

Discussion: https://github.com/graphql/graphql-wg/discussions/1700

> This proposal introduces a directive that can be added to fields to indicate that they should propagate errors.

Reviewer: ```suggestion
This proposal changes the `!` symbol and the `NON_NULL` introspection kind both to mean "semantic non null" (allowing for `null` on error). It introduces the `@propagateError` directive that can be added to fields to indicate that they should propagate errors in order to provide backwards compatibility with existing deployed clients.
```
